### ISSUE TITLE
Implement "star merge" functionality

### ISSUE THREAD

**Issue by bdbaddog** [SCons/scons#1905]
This issue was originally created at: 2008-02-09 05:06:00.
This issue was reported by: `gregnoel`.
gregnoel said at 2008-02-09 05:06:00
>If the same repository is used for all files and directories within some subtree (that is, a `SourceCode('.', xxx)` was given for some directory and never changed below it), then only issue one call to the repository to refresh the entire directory tree.  This is the simple case that will cause 99.44% of repository fetches to be a single operation.
> 
>For the more complex cases where repository sources are intermixed, surely some of fragments can be combined.  In the worst case, fall back to file-by-file access.
> 
>I'd request that the simpler case in the first paragraph be implemented as soon as possible; the more complex case can be delayed until there's some convergence on the API.

gregnoel said at 2008-02-19 15:13:04
>I'm having second thoughts about this.  I think I was too glib with my initial assessment.
> 
>Let's consider the case where the variant, working, and source directories are all different.  In this case, you want to do a full checkout in _source_ directories, a file-by-file checkout in working directories, and no checkout in variant directories.  The rationale is that the full-directory fetch in the repository gets all the updates in a single operation; much less load on a remote repository.  The file-by-file fetch in the working directories picks up changes to files that the user is working on; there are usually only a few files, relative to the number of files in the full checkout.  If there's a built file that's also in the checkout (including a foo.c file derived from a foo.y file for people who didn't have yacc used to be a common paradigm) you don't want to update your built file (mostly because it'll probably fail) and a routine rebuild will cause the file to be regenerated anyway.
> 
>The interesting cases are where the directory trees overlap.
> 
>If the source and working directories are the same, you still want to do a full checkout to reduce server overhead and you don't want to do file-by-file checkouts since the full checkout would have taken care of it.
> 
>If the working and variant directories are the same, you still want to do a file-by-file checkout for leaf files but suppress checkouts for built files.
> 
>The most difficult case is when all three directory trees are the same.  On the one hand, you want to do a full checkout in one operation and not do file-by-file checkouts.  On the other hand, you want to suppress checkouts for built targets so that spurious upstream changes don't cause errors.  I'm pretty sure that neither SCS nor SVN can have exclusion lists, so the only way to deal with it seems to be to identify when built files have conflicts and ignore those errors, something that sounds awfully hard to do in general.
> 
>What is really needed initially is an API that will cause a full checkout into the source directories.  Since that could be chancy if there are more than one set of source directories (which one to update?), perhaps a starting point would be an API that would do a full checkout if the source and working directories are the same, since this case is the most common one.
> 
>Some way of doing a full checkout as part of the normal processing probably needs to be in place prior to 1.0.

stevenknight said at 2008-04-10 01:59:23
>Target milestone 1.0 per spreadsheet consensus.

stevenknight said at 2008-04-11 04:57:50
>Put 1.0 target milestone on wrong issue, resetting to -unspecified-.

gregnoel said at 2008-04-14 12:57:25
>Bug party triage: Steven to float a message on the mailing lists about retiring SourceCode() (and CVS(), SCCS(), RCS(), Perforce(), Subversion(), et.al.).

stevenknight said at 2009-02-25 07:12:16
>I polled the mailing list back in November (and then didn't follow up and update this issue).  Here's the original message:
> 
>http://scons.tigris.org/ds/viewMessage.do?dsForumId=1272&dsMessageId=92706
> 
>Unfortunately, tigris.org didn't thread the replies, so you need to search for SourceCode in the subject to see the ones that went to the ML.  A few people replied to me off-line.
> 
>Of the 16 respondents, only one used SourceCode():
> 
>>>>>
>    I vote for keeping the SourceCode() feature. I think until a solution is found, it is the only thing we have. It is better than nothing.
> 
>    This feature is absolutely crucial for component based builds, where each components brings in their dependency from the repository or for that matter from anywhere else. It seems to me people have done a lot of acrobatics to get this requirement implemented using other methods. A lot of acrobatics indicates usually a real unmet need. If SCons solves it in its own context, so much the better.
><<<<
> 
>There was some sentiment from people saying they'd *like* to use it, but tried and found that it was not scalable, or didn't work as they expected when things like VariantDir are configured.
> 
>It sounds like we can make a pretty good case for getting rid of it, at least in its current form.  It also sounds like the real underlying issue here is that people might *like* some way for the build system to be able to fetch components for building automatically, but SourceCode() definitely isn't it, and if we were to design a good mechanism for this it might be really valuable.
> 
>(Note on this score that Bram Moolenaar's A-A-P build tool had a model for fetching Open Source software components for building that some people found pretty compelling.  And it was in Python, too, so there might be some stuff to borrow there.)
> 
>I'm changing the Target milestone back to -unspecified- so we can discuss these issues at the next Bug party.

gregnoel said at 2009-03-03 17:00:13
>I agree with Steven, we should remove SourceCode().  I propose to replace the functionality with two other features: the ability to run commands before walking some part of the source tree (already in the TaskmasterNG design) and the ability to do a star-merge from a repository into a file in the working directory.
> 
>The former capability could be used to do a checkout in any variation the user desired.  We'd document some of the obvious examples.  Here's one, assuming that 'Sequential()' is the API to get the before/after functionality:
```python
    Command(Alias('checkout'), [], 'svn update')
    AlwaysBuild(Alias('checkout'))
    Sequential(Alias('checkout'), '.')
```
>This idiom would update the local directory before starting to process it.  Or if the source tree is to be kept pristine:
```python
    rep = Dir('/some/path/to/a/directory')
    Repository(dir)
    Command(dir, [], 'svn checkout ... $TARGET')
    AlwaysBuild(dir)
    Sequential(dir, '.')
```
>I haven't tried a repository that's also a build target, but there's no obvious reason why it shouldn't work.
> 
>The idea behind the star-merge is suggested by the second example.  Note that if a new source file shows up and it's also being edited in the working directory, the changes aren't merged into the working directory.  That's what the star-merge does.  At least two Python implementations (Hg and Bzr) already exist; we may be able to steal one of them.  One would then say
```python
    StarMerge('src')
```
>to indicate that if there's a local file and the upstream file changes, the changes should be merged into the local file.  Exactly how that should happen will need to be worked out, but it should be independent of how changes are made to the upstream file.  This will presumably require SCons to keep track of the previous contents of the upstream file.

gregnoel said at 2009-03-05 10:41:31
>Bug party triage.  Spin off an issue to deprecate SourceCode() and its related functions.  Spin off an issue to document patterns that will do a VCS checkout into a directory.  Make this issue be about implementing a "star-merge" feature so a locally-modified file will be updated if the upstream source changes.

garyo said at 2012-09-01 10:04:01
>de-assigning all tickets assigned to Greg Noel (no longer working on SCons)

**Comment by mwichmann** [comment]
Hmmm, SourceCode is gone, but I'm not sure whether that means this can all go away...

**Comment by bdbaddog** [comment]
Closing. Won't implement.
